src/backend/d3d12/CommandBufferD3D12.cpp

```diff
@@ -226,130 +226,54 @@ namespace backend { namespace d3d12 {
                     D3D12_DESCRIPTOR_HEAP_TYPE_SAMPLER);
             }
         }
-    }  // namespace
 
-    CommandBuffer::CommandBuffer(Device* device, CommandBufferBuilder* builder)
-        : CommandBufferBase(builder), mDevice(device), mCommands(builder->AcquireCommands()) {
+    }  // anonymous namespace
+
+    CommandBuffer::CommandBuffer(CommandBufferBuilder* builder)
+        : CommandBufferBase(builder), mCommands(builder->AcquireCommands()) {
     }
 
     CommandBuffer::~CommandBuffer() {
         FreeCommands(&mCommands);
     }
 
-    void CommandBuffer::FillCommands(ComPtr<ID3D12GraphicsCommandList> commandList) {
-        BindGroupStateTracker bindingTracker(mDevice);
-        AllocateAndSetDescriptorHeaps(mDevice, &bindingTracker, &mCommands);
-        bindingTracker.Reset();
-
-        ID3D12DescriptorHeap* descriptorHeaps[2] = {bindingTracker.cbvSrvUavGPUDescriptorHeap.Get(),
-                                                    bindingTracker.samplerGPUDescriptorHeap.Get()};
-        if (descriptorHeaps[0] && descriptorHeaps[1]) {
-            commandList->SetDescriptorHeaps(2, descriptorHeaps);
-        } else if (descriptorHeaps[0]) {
-            commandList->SetDescriptorHeaps(1, descriptorHeaps);
-        } else if (descriptorHeaps[1]) {
-            commandList->SetDescriptorHeaps(2, &descriptorHeaps[1]);
+    void CommandBuffer::RecordCommands(ComPtr<ID3D12GraphicsCommandList> commandList) {
+        Device* device = ToBackend(GetDevice());
+        BindGroupStateTracker bindingTracker(device);
+
+        // Precompute the allocation of bindgroups in descriptor heaps
+        // TODO(cwallez@chromium.org): Iterating over all the commands here is inefficient. We
+        // should have a system where commands and descriptors are recorded in parallel then the
+        // heaps set using a small CommandList inserted just before the main CommandList.
+        {
+            AllocateAndSetDescriptorHeaps(device, &bindingTracker, &mCommands);
+            bindingTracker.Reset();
+
+            ID3D12DescriptorHeap* descriptorHeaps[2] = {
+                bindingTracker.cbvSrvUavGPUDescriptorHeap.Get(),
+                bindingTracker.samplerGPUDescriptorHeap.Get()};
+            if (descriptorHeaps[0] && descriptorHeaps[1]) {
+                commandList->SetDescriptorHeaps(2, descriptorHeaps);
+            } else if (descriptorHeaps[0]) {
+                commandList->SetDescriptorHeaps(1, descriptorHeaps);
+            } else if (descriptorHeaps[1]) {
+                commandList->SetDescriptorHeaps(2, &descriptorHeaps[1]);
```

Comment:
Not new, but is this valid? Isn't this saying "array of 2 descriptor heaps starting at index 1" (in a size-2 array)?

Reply:
Done